test/functional/wallet_upgradewallet.py

```diff
@@ -0,0 +1,125 @@
+#!/usr/bin/env python3
+# Copyright (c) 2018-2020 The Bitcoin Core developers
+# Distributed under the MIT software license, see the accompanying
+# file COPYING or http://www.opensource.org/licenses/mit-license.php.
+"""upgradewallet RPC functional test
+
+Test upgradewallet RPC. Download v0.15.2 v0.16.3 node binaries:
+
+contrib/devtools/previous_release.sh -b v0.15.2 v0.16.3
+"""
+
+import os
+import shutil
+
+from test_framework.test_framework import BitcoinTestFramework, SkipTest
+from test_framework.util import (
+    adjust_bitcoin_conf_for_pre_17,
+    assert_equal,
+    assert_greater_than
+)
+
+class UpgradeWalletTest(BitcoinTestFramework):
+    def set_test_params(self):
+        self.setup_clean_chain = True
+        self.num_nodes = 3
+        self.extra_args = [
+            ["-addresstype=bech32"], # current wallet version
+            ["-usehd=1"],            # v0.16.3 wallet
+            ["-usehd=0"]             # v0.15.2 wallet
+        ]
+
+    def skip_test_if_missing_module(self):
+        self.skip_if_no_wallet()
+
+    def setup_network(self):
+        self.setup_nodes()
+
+    def setup_nodes(self):
+        if os.getenv("TEST_PREVIOUS_RELEASES") == "false":
+            raise SkipTest("upgradewallet RPC tests")
+
+        releases_path = os.getenv("PREVIOUS_RELEASES_DIR") or os.getcwd() + "/releases"
+        if not os.path.isdir(releases_path):
+            if os.getenv("TEST_PREVIOUS_RELEASES") == "true":
+                raise AssertionError("TEST_PREVIOUS_RELEASES=1 but releases missing: " + releases_path)
+            raise SkipTest("This test requires binaries for previous releases")
+
+        self.add_nodes(self.num_nodes, extra_args=self.extra_args, versions=[
+            None,
+            160300,
+            150200
+        ], binary=[
+            self.options.bitcoind,
+            releases_path + "/v0.16.3/bin/bitcoind",
+            releases_path + "/v0.15.2/bin/bitcoind",
+        ], binary_cli=[
+            self.options.bitcoincli,
+            releases_path + "/v0.16.3/bin/bitcoin-cli",
+            releases_path + "/v0.15.2/bin/bitcoin-cli",
+        ])
+        # adapt bitcoin.conf, because older bitcoind's don't recognize config sections
+        adjust_bitcoin_conf_for_pre_17(self.nodes[1].bitcoinconf)
+        adjust_bitcoin_conf_for_pre_17(self.nodes[2].bitcoinconf)
+        self.start_nodes()
+
+    def run_test(self):
+        self.nodes[0].generatetoaddress(101, self.nodes[0].getnewaddress())
+        # # Sanity check the test framework:
+        res = self.nodes[0].getblockchaininfo()
+        assert_equal(res['blocks'], 101)
+        node_master = self.nodes[0]
+        v16_3_node  = self.nodes[1]
+        v15_2_node  = self.nodes[2]
+
+        self.log.info("Test upgradewallet RPC...")
```

Comment:
It would be nice to send some coins around, so that we can check the balance doesn't decrease from upgrading the wallet

Reply:
Sadly, this can't be implemented, because old nodes aren't connected to *node_master*. We had this problem with rpc_connections previously when I tried to filter out old nodes. Now, the same problem happens when trying to use *sync_* methods.

Reply:
More precisely, this is the place where the error gets thrown.

```shell
 File "/test/functional/test_framework/test_framework.py", line 499, in sync_blocks
    sync_blocks(nodes or self.nodes, **kwargs)
  File "/test/functional/test_framework/util.py", line 418, in sync_blocks
    assert (all([len(x.getpeerinfo()) for x in rpc_connections]))
```

We have seen this already. So, the question is, do we want to work on "onboarding" of old nodes so their converted wallets can be tested here, or should we rather open a dedicated PR for this?

Reply:
So syncing the blockchain is the issue?

Can be done in a really dumb way:

```
def dumb_sync_blocks(...):
 node_from = select node with max height
 for i in range(max(blockheight)):
  b = node_from.getblock(lbockhash=node_from.getblockhash(i), verbose=0)
  for n in nodes:
    n.submitblock(b)

Reply:
Thanks. Have tried it, but the v0.15.2 node doesn't expand its chain with submitted blocks. Interestingly, in its blockchaininfo the header count gets increased while block count remains at 0. The other node, v0.16.3 is working fine. Here my variant of *dumb_sync_blocks*

```python
def dumb_sync_blocks(self):
    node_from = self.nodes[0]
    to_height = node_from.getblockheader(node_from.getbestblockhash())['height']
    for n in range(1, len(self.nodes)):
        height = self.nodes[n].getblockheader(self.nodes[n].getbestblockhash())['height']
        for i in range(height, to_height+1):
            b = node_from.getblock(blockhash=node_from.getblockhash(i), verbose=0)
            self.nodes[n].submitblock(b)
        new_height = self.nodes[n].getblockheader(self.nodes[n].getbestblockhash())['height']
        print("node: {}, new best height: {}".format(n, new_height))
```

The output is like this:

```shell
node: 1, new best height: 101
node: 2, new best height: 0
```

Here their respective blockchaininfo jsons:

**v0.16.3**
```js
{
'chain': 'regtest', 
'blocks': 101, 
'headers': 101, 
'bestblockhash': '1a2f4021b111f20b12071bb68c92dc0037a2237ae034b345b002971dbb9f5e16', }
```

**v0.15.2**
```js
{
'chain': 'regtest', 
'blocks': 0, 
'headers': 101, 
'bestblockhash': '0f9188f13cb7b2c71f2a335e3a4fc328bf5beb436012afca590b1a11466e2206' }
```

In the test code I am sending some coins to both wallets and then calling *dumb_sync_blocks* to update all chains.

```python
node_master.generatetoaddress(101, v16_3_address)
self.dumb_sync_blocks()
```

Reply:
Ah, right. regtest was hardforked

Reply:
Ok, then I will be doing balance tests with v0.16.3 only? v0.15.2 can still be used for upgrade tests. Is this acceptable?

Reply:
It might be possible to apply a hack and either:

* never produce witnesses (use legacy-1-addresses)
* maybe strip the witness from the blocks before submitting, but I doubt that works because VERIFY_WITNESS should be enforced from genesis, always.

Reply:
You don't have to do this, I am happy to merge the pull in its current state. Just let me know.

Reply:
Thanks. I think I should try to implement it. If it works, fine. If not, then you can merge the current state. Will come back to you later.

Reply:
Well, I have managed to sync the chains, but there is no way to make proper transactions. I have stolen a few things from *p2p_segwit*, like *build_next_block* for manual block generation and syncing:

```js
node: 1, new best height: 101
node: 2, new best height: 101
{
'chain': 'regtest', 
'blocks': 101, 
'headers': 101, 
'bestblockhash': '0f2cfae3d63464c070eea6a8305f6f02245e69f41987c186da1c7f8af2bb1547', }
{
'chain': 'regtest', 
'blocks': 101, 
'headers': 101, 
'bestblockhash': '0f2cfae3d63464c070eea6a8305f6f02245e69f41987c186da1c7f8af2bb1547'}
``` 

However, I see no way how to update balance of v0.15.2 as I'd need to build a transaction that would have to have the witness field.

I will now push the latest changes.